plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -935,9 +941,83 @@ public Optional<ConnectorTableHandle> applyDelete(ConnectorSession session, Conn
     }
 
     @Override
-    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle)
+    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        if (table.getFormatVersion() < 2) {
+            throw new TrinoException(NOT_SUPPORTED, "Iceberg table updates require at least format version 2");
+        }
+        String deleteMode = table.getStorageProperties().get(DELETE_MODE);
+        if (!"merge-on-read".equals(deleteMode)) {
+            String errorMessage = "Row level deletes are only supported for tables with '" + DELETE_MODE + "' set to 'merge-on-read'.";
+            if (deleteMode != null) {
+                errorMessage += " '" + deleteMode + "' is not supported.";
+            }
+            throw new TrinoException(NOT_SUPPORTED, errorMessage);
+        }
+        verify(transaction == null, "transaction already set");
+        transaction = catalog.loadTable(session, table.getSchemaTableName()).newTransaction();
+        return table.withRetryMode(retryMode);
+    }
+
+    @Override
+    public void finishDelete(ConnectorSession session, ConnectorTableHandle tableHandle, Collection<Slice> fragments)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        Table icebergTable = transaction.table();
```

Comment:
This is not really related to this change but I've never paid attention to this. 
Is it thread safe to store transaction in a field ? Isn't IcebergMetadata reused?

Reply:
It is, each transaction gets it's own `IcebergMetadata`. You can see that in `IcebergTransactionManager`